### PR TITLE
Removed Thrift service hierarchy in preparation for service changes

### PR DESCRIPTION
**Pull request by dlmarion**
This change modifies TabletClientService such that it no longer extends
ClientService and ManagerClientService such that it no longer extends
FateService. This is in preparation for splitting TabletClientService
into at least two services for the ScanServer feature. The TabletServer
and Manager processes needed to be modified to start two Thrift Processors
instead of one. This required the use of the TMultiplexedProcessor object
on the server side, which takes one or more Processors and places them in
a map keyed on the Thrift service name. Clients, when communicating with
the TabletServer or Manager, then need to use a TMultiplexedProtocol with
the name of the Thrift service that they want to call.

### CODE REVIEW

**Comment by dlmarion** [comment]
Full IT build passed: https://github.com/dlmarion/accumulo/actions/runs/2116766533

File: server/manager/src/main/java/org/apache/accumulo/manager/ManagerClientServiceHandler.java [hunk 1]
```diff
@@ -96,15 +96,15 @@
...
 import com.google.protobuf.InvalidProtocolBufferException;
 
-public class ManagerClientServiceHandler extends FateServiceHandler
-    implements ManagerClientService.Iface {
+public class ManagerClientServiceHandler implements ManagerClientService.Iface {
 
   private static final Logger log = Manager.log;
   private static final Logger drainLog =
       LoggerFactory.getLogger("org.apache.accumulo.manager.ManagerDrainImpl");
+  protected final Manager manager;
```
Comment (ctubbsii) [inline server/manager/src/main/java/org/apache/accumulo/manager/ManagerClientServiceHandler.java:104]: I don't think this needs to be protected anymore. Can probably be private now, or package-private. Same with the one in FateServiceHandler.java.

**Review by ctubbsii — COMMENTED** [review]
I like the idea of using a multiplexed processor to handle the different thrift service scopes, so we can create a service that is built using smaller service building blocks. I wasn't aware that a multiplexed processor existed before. The title of this PR and description I think could be improved... it's not a problem in the PR, but the git commit when it gets merged should be updated to make it more obvious that the main thing this change does is make our service definitions more modular by switching to a multiplexed Thrift protocol.

Are there any performance implications for switching to this multiplexed protocol?

Overall, I'm in favor of this change. I do worry that change to Thrift code tend to break in surprising ways, but I think it will be fine with sufficient testing, as this is nothing new when working with Thrift.

My main concern right now is that maybe we need to improve some of the names of these classes / utility methods. Some of them are not very indicative of what they are supposed to do.

Also, I think this multiplexed stuff completely breaks all our RPC layers. Any changes we have been hesitant to do for any reason, are now on the table with move to a multiplexed protocol, I think, such as stripping out any deprecated fields we previously left in, or making all fields optional, further breaking up the methods into discrete service definitions, etc. If everything is broken anyway, now is the best time to make any other breaking changes, so that way we might minimize breaking changes in future.

File: core/src/main/thrift-gen-java/org/apache/accumulo/core/manager/thrift/ManagerClientService.java [hunk 2]
```diff
@@ -27,7 +27,7 @@
 @SuppressWarnings({"cast", "rawtypes", "serial", "unchecked", "unused"})
 public class ManagerClientService {
 
-  public interface Iface extends FateService.Iface {
+  public interface Iface {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/thrift-gen-java/org/apache/accumulo/core/manager/thrift/ManagerClientService.java:30]: Correct, the thrift IDL was changed and the Thrift code regenerated.

**Comment by dlmarion** [comment]
> so we can create a service that is built using smaller service building blocks.

Looking at how this could be broken down, I'm wondering if it makes sense to group the Thrift calls made in support of client operations (Instance, Namespace, Replication, Security, Table) into Thrift Service definitions (InstanceService, NamespaceService, ...). For the tablet mutating objects (BatchWriter, BatchDeleter, ...) create a TabletWriterService, and a TabletReaderService for tablet scanning objects (Scanner, BatchScanner, ...)

I'm thinking that this could be a follow-on task to this PR.

File: server/manager/src/main/java/org/apache/accumulo/manager/ManagerClientServiceHandler.java (hunk 1 above)
Comment (dlmarion) [reply server/manager/src/main/java/org/apache/accumulo/manager/ManagerClientServiceHandler.java]: Addressed in e243d49 and c0814db

File: core/src/main/java/org/apache/accumulo/core/clientImpl/FateClient.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import org.apache.accumulo.core.manager.thrift.FateService;
+import org.apache.accumulo.core.rpc.ThriftUtil;
+import org.apache.accumulo.core.trace.TraceUtil;
+import org.apache.accumulo.core.util.HostAndPort;
+import org.apache.thrift.TServiceClient;
+import org.apache.thrift.transport.TTransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FateClient {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/FateClient.java]: Addressed in e243d49, renamed class to `FateManagerClient`

File: core/src/main/java/org/apache/accumulo/core/rpc/ThriftUtil.java [hunk 4]
```diff
@@ -163,7 +179,35 @@ public static void returnClient(TServiceClient iface, ClientContext context) {
...
   /**
-   * Create a TabletServer Thrift client
+   * Create a TabletServer ClientService client
+   *
+   * @param address
+   *          Server address for client to connect to
+   * @param context
+   *          RPC options
+   */
+  public static ClientService.Client getClientServiceClient(HostAndPort address,
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/rpc/ThriftUtil.java]: Addressed in e243d49, I ended up removing the `get*Client` methods and modified callers to use `getClient` method.

File: test/src/main/java/org/apache/accumulo/test/performance/NullTserver.java [hunk 5]
```diff
@@ -92,14 +91,11 @@
...
-  public static class ThriftClientHandler extends ClientServiceHandler
-      implements TabletClientService.Iface {
+  public static class ThriftClientHandler implements TabletClientService.Iface {
 
     private long updateSession = 1;
 
-    public ThriftClientHandler(ServerContext context, TransactionWatcher watcher) {
-      super(context, watcher);
-    }
+    public ThriftClientHandler(ServerContext context, TransactionWatcher watcher) {}
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/performance/NullTserver.java]: Addressed in e243d49, constructor removed

**Comment by dlmarion** [comment]
I noticed while working on this that ManagerClient, FateManagerClient, ReplicationClient and ServerClient have a good amount of code duplication. I think it might be possible to consolidate them. Not sure if that should be part of this PR or another issue.

**Review by keith-turner — COMMENTED** [review]
> Overall, I'm in favor of this change. I do worry that change to Thrift code tend to break in surprising ways, but I think it will be fine with sufficient testing, as this is nothing new when working with Thrift.

I am curious what the behavior is when a 2.1 client connects to a 2.0 server or visa versa a 2.0 client connects to a 2.1 server with this changes.  Ideally there would be a nice error message that indicates RPC incompatibility, not sure if thrift will do this though.

> Are there any performance implications for switching to this multiplexed protocol?

I am also curious about this.

**Comment by dlmarion** [comment]
> > Overall, I'm in favor of this change. I do worry that change to Thrift code tend to break in surprising ways, but I think it will be fine with sufficient testing, as this is nothing new when working with Thrift.
> 
> I am curious what the behavior is when a 2.1 client connects to a 2.0 server or visa versa a 2.0 client connects to a 2.1 server with this changes. Ideally there would be a nice error message that indicates RPC incompatibility, not sure if thrift will do this though.
> 

 It definitely is the case that if you have attempt to use a non-multiplexed client with a multiplexed server, then you get a message in the error suggesting that you should be using the multiplexed client. The error that is returned to the client is at https://github.com/apache/thrift/blob/master/lib/java/src/org/apache/thrift/TMultiplexedProcessor.java#L116

> > Are there any performance implications for switching to this multiplexed protocol?
> 
> I am also curious about this.

I did look for issues in the Thrift JIRA regarding the multiplexed processors, but I didn't really find much. The client passes the service name and the message to the server, and in [this](https://github.com/apache/thrift/blob/master/lib/java/src/org/apache/thrift/TMultiplexedProcessor.java#L79) method the server uses the service name to get the correct Processor from a map and then passes the message to that Processor. I think there will be a small amount of time taken per API call to perform the logic in the TMultiplexedProcessor.process method. Do you think I should benchmark it?

File: server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java [hunk 6]
```diff
@@ -162,17 +164,22 @@
 import io.opentelemetry.api.trace.Span;
 import io.opentelemetry.context.Scope;
 
-public class ThriftClientHandler extends ClientServiceHandler implements TabletClientService.Iface {
+public class ThriftClientHandler implements TabletClientService.Iface {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java]: Renamed to `TabletClientHandler` in d582824abc7df65759f6ebe2dc3d55e0ec554daa

File: server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java [hunk 7]
```diff
@@ -162,17 +164,22 @@
...
+  protected final TransactionWatcher transactionWatcher;
+  protected final ServerContext context;
+  protected final SecurityOperation security;
   private final WriteTracker writeTracker = new WriteTracker();
   private final RowLocks rowLocks = new RowLocks();
 
   public ThriftClientHandler(TabletServer server) {
-    super(server.getContext(), new TransactionWatcher(server.getContext()));
+    this.context = server.getContext();
+    this.transactionWatcher = new TransactionWatcher(server.getContext());
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java]: Modified to create a TransactionWatcher in TabletServer and pass the instance to the ClientHandler and the TabletClientHandler in d582824abc7df65759f6ebe2dc3d55e0ec554daa

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java [hunk 8]
```diff
@@ -600,30 +604,32 @@ private void returnManagerConnection(ManagerClientService.Client client) {
...
-    Iface rpcProxy = TraceUtil.wrapService(clientHandler);
-    final Processor<Iface> processor;
-    if (getContext().getThriftServerType() == ThriftServerType.SASL) {
-      Iface tcredProxy = TCredentialsUpdatingWrapper.service(rpcProxy, ThriftClientHandler.class,
-          getConfiguration());
-      processor = new Processor<>(tcredProxy);
-    } else {
-      processor = new Processor<>(rpcProxy);
+    clientHandler = getClientHandler();
+    thriftClientHandler = getThriftClientHandler();
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java]: Implemented suggestion in d582824abc7df65759f6ebe2dc3d55e0ec554daa

File: server/base/src/main/java/org/apache/accumulo/server/rpc/ThriftServerTypes.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+import org.apache.accumulo.core.trace.TraceUtil;
+import org.apache.accumulo.server.ServerContext;
+import org.apache.accumulo.server.client.ClientServiceHandler;
+import org.apache.thrift.TBaseProcessor;
+import org.apache.thrift.TMultiplexedProcessor;
+import org.apache.thrift.TProcessor;
+import org.apache.thrift.TServiceClient;
+import org.apache.thrift.TServiceClientFactory;
+
+public class ThriftServerTypes {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/ThriftServerTypes.java]: Renamed `ThriftServerTypes` to `ThriftProcessorTypes` in d582824abc7df65759f6ebe2dc3d55e0ec554daa

File: server/base/src/main/java/org/apache/accumulo/server/rpc/ThriftServerTypes.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+      TabletClientService.Client.Factory> TABLET_SERVER =
+          new ServerType<>(ThriftClientTypes.TABLET_SERVER);
+
+  public static TProcessor getCompactorThriftServer(CompactorService.Iface serviceHandler,
+      ServerContext context, AccumuloConfiguration conf) throws Exception {
+    return COMPACTOR.getServer(CompactorService.Processor.class, CompactorService.Iface.class,
+        serviceHandler, context, conf);
+  }
+
+  public static TProcessor getCoordinatorThriftServer(
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/ThriftServerTypes.java]: Renamed methods to `get{}TProcessor()` in d582824abc7df65759f6ebe2dc3d55e0ec554daa

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java [hunk 11]
```diff
@@ -600,30 +604,32 @@ private void returnManagerConnection(ManagerClientService.Client client) {
...
   private HostAndPort startTabletClientService() throws UnknownHostException {
     // start listening for client connection last
-    clientHandler = getThriftClientHandler();
-    Iface rpcProxy = TraceUtil.wrapService(clientHandler);
-    final Processor<Iface> processor;
-    if (getContext().getThriftServerType() == ThriftServerType.SASL) {
-      Iface tcredProxy = TCredentialsUpdatingWrapper.service(rpcProxy, ThriftClientHandler.class,
-          getConfiguration());
-      processor = new Processor<>(tcredProxy);
-    } else {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:610]: The logic moved to `ThriftProcessorTypes.ProcessorType.getTProcessor()`

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java [hunk 12]
```diff
@@ -582,13 +582,17 @@ private ManagerClientService.Client managerConnection(HostAndPort address) {
...
       // log.info("Listener API to manager has been opened");
-      return ThriftUtil.getClient(new ManagerClientService.Client.Factory(), address, getContext());
+      return ThriftUtil.getClient(ThriftClientTypes.MANAGER, address, getContext());
     } catch (Exception e) {
       log.warn("Issue with managerConnection (" + address + ") " + e, e);
     }
     return null;
   }
 
+  protected ClientServiceHandler getClientHandler() {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java]: renamed this and the other related method in 5df51dd9da8c4b48b5913cfbcac7e0cd48d62ef7

**Comment by dlmarion** [comment]
> > Do you think I should benchmark it?
> 
> If you want to. A map lookup seems really trivial compared to the overhead of the IO

That's what I'm thinking. I'm thinking the overhead is negligible. The map is created when the TProcessor is set up and is static. The Thrift server thread comes into the process method and decodes the message to know which TProcessor implementation to send it to.

**Comment by ctubbsii** [comment]
Okay. I'll try to take a more thorough look at this soon. I want to prioritize looking @EdColeman 's #2569, though.

**Comment by dlmarion** [comment]
Mostly successful IT run (1 test failed due to ZK timeout): https://github.com/dlmarion/accumulo/actions/runs/2189701856

**Comment by dlmarion** [comment]
> I am curious what the behavior is when a 2.1 client connects to a 2.0 server or visa versa a 2.0 client connects to a 2.1 server with this changes.

I think they fail in both scenarios. I was unable to find the location where an error is from the server side when a multiplexed protocol is used by the client (2.1) but not by the server (2.0). Thrift does return a good error message when a non-multiplexed client is used against a multiplexed server (see above). This would mean that users would need to upgrade their clients to use the 2.1 libraries when they upgrade to 2.1 on the server side.

**Review by keith-turner — APPROVED** [review]

**Comment by dlmarion** [comment]
> It does add a small amount of data to each RPC request, looking at the javadoc for `TMultiplexedProcessor` it seems the servicename from the thrift IDL is added to each RPC request. So scan RPCs may now have the extra data `TabletClientService` from the service name defined in the idl. Not sure if this will matter or if it would be worthwhile to have a shorter service name.

I know the javadoc says that the service name from the Thrift IDL is to be used, but I don't see anything in the code that enforces it. I think it might be possible to leave the IDL as-is and just register the TProcessors with a different service name in the `TMultiplexedProcessor.register(String serviceName, TProcessor processor)` [call](https://github.com/apache/thrift/blob/master/lib/java/src/org/apache/thrift/TMultiplexedProcessor.java#L67). If that works, then we could have an enum or something that is used by the client and server.

File: core/src/main/java/org/apache/accumulo/core/rpc/ThriftUtil.java [hunk 13]
```diff
@@ -91,63 +92,67 @@ public static TTransportFactory transportFactory() {
...
    * Create a Thrift client using the given factory and transport
    */
-  public static <T extends TServiceClient> T createClient(TServiceClientFactory<T> factory,
+  public static <T extends TServiceClient> T createClient(ThriftClientType<T,?> type,
       TTransport transport) {
-    return factory.getClient(protocolFactory.getProtocol(transport),
-        protocolFactory.getProtocol(transport));
+
+    TProtocol protocol = protocolFactory.getProtocol(transport);
+    if (type.isMultiplexed()) {
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/rpc/ThriftUtil.java:99]: I think this can be simplified slightly if we just make everything multiplexed (since it's basically no cost, and can help make rpc extensible in future).

**Comment by keith-turner** [comment]
> I know the javadoc says that the service name from the Thrift IDL is to be used, but I don't see anything in the code that enforces it. I think it might be possible to leave the IDL as-is and just register the TProcessors with a different service name in the TMultiplexedProcessor.register(String serviceName, TProcessor processor) [call](https://github.com/apache/thrift/blob/master/lib/java/src/org/apache/thrift/TMultiplexedProcessor.java#L67). If that works, then we could have an enum or something that is used by the client and server.

Oh interesting that its just a convention.  Yeah it may be worthwhile to map all the service name to short two or three letter codes.  It seems like the way you have organized the code in the PR that that mapping could be isolated to one or two places i the code so it would be well encapsulated.

File: server/manager/src/main/java/org/apache/accumulo/manager/Manager.java [hunk 14]
```diff
@@ -1014,30 +1014,32 @@ public void run() {
...
+            processor, "Manager", "Manager Client Service Handler", null,
+            Property.MANAGER_MINTHREADS, Property.MANAGER_MINTHREADS_TIMEOUT,
+            Property.MANAGER_THREADCHECK, Property.GENERAL_MAX_MESSAGE_SIZE);
+      } catch (UnknownHostException e) {
+        throw new IllegalStateException("Unable to start server on host " + getHostname(), e);
+      }
+      clientService = sa.server;
+      log.info("Started Manager client service at {}", sa.address);
+    } catch (Exception e2) {
+      throw new RuntimeException("Error creating thrift server processor", e2);
```
Comment (ctubbsii) [inline server/manager/src/main/java/org/apache/accumulo/manager/Manager.java:1041]: Can this catch something more specific?

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java [hunk 15]
```diff
@@ -600,30 +604,33 @@ private void returnManagerConnection(ManagerClientService.Client client) {
...
+    thriftClientHandler = newTabletClientHandler(watcher);
+
+    try {
+      TProcessor processor = ThriftProcessorTypes.getTabletServerTProcessor(clientHandler,
+          thriftClientHandler, getContext(), getConfiguration());
+      HostAndPort address = startServer(getConfiguration(), clientAddress.getHost(), processor);
+      log.info("address = {}", address);
+      return address;
+    } catch (Exception e) {
+      throw new RuntimeException("Error creating thrift server processor", e);
```
Comment (ctubbsii) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:618]: Can this be a more specific exception?

File: server/manager/src/main/java/org/apache/accumulo/manager/Manager.java (hunk 14 above)
Comment (dlmarion) [reply server/manager/src/main/java/org/apache/accumulo/manager/Manager.java]: I was able to remove the Exception declaration from `ThriftProcessorTypes.ProcessorType.getProcessor()` and instead throw an `IllegalArgumentException`. This allowed me to remove this completely in 2bac8d5f30828ada41d60168272b0c6b46802dc5

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java (hunk 15 above)
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java]: I was able to remove the Exception declaration from `ThriftProcessorTypes.ProcessorType.getProcessor()` and instead throw an `IllegalArgumentException`. This allowed me to remove this completely in 2bac8d5f30828ada41d60168272b0c6b46802dc5

File: core/src/main/java/org/apache/accumulo/core/rpc/ThriftUtil.java (hunk 13 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/rpc/ThriftUtil.java]: I made everything multiplexed in 2bac8d5f30828ada41d60168272b0c6b46802dc5

**Comment by ctubbsii** [comment]
@dlmarion Just a heads up that I saw LateLastContactIT failing due to timeouts beginning when this was merged. I don't know if it's related, and haven't had time to troubleshoot, but just wanted to mention it here, in case you saw it too and maybe could reproduce it. We talked about performance impact above, but I doubt this had any significant impact on performance, since the overall build time hasn't really changed... it's just this test. So, maybe there's an edge case with some thrift stuff with this one? Or, could be entirely unrelated. :shrug:

**Comment by dlmarion** [comment]
I did notice it, I commented on it at https://github.com/apache/accumulo/pull/2643. I'll see if I can figure it out.

**Comment by dlmarion** [comment]
Merged a fix for it - #2645
